content/posts/gateway-vs-ingress-api.md

```diff
@@ -0,0 +1,266 @@
+kind: Ingress
+metadata:
+  name: api-routes
+spec:
+  ingressClassName: apisix
+  rules:
+    - host: local.navendu.me
+      http:
+        paths:
+          - backend:
+              service:
+                name: bare-minimum-api-v1
+                port:
+                  number: 8080
+            path: /v1
+            pathType: Prefix
+          - backend:
+              service:
+                name: bare-minimum-api-v2
+                port:
+                  number: 8081
+            path: /v2
+            pathType: Prefix
+```
+
+> **Tip**: You can check out this [hands-on tutorial](/posts/hands-on-set-up-ingress-on-kubernetes-with-apache-apisix-ingress-controller/) to learn more about setting up Ingress on Kubernetes with Apache APISIX Ingress controller.
+
+Since the Ingress API is not tied to any particular controller implementation, you can swap APISIX with any other Ingress controller, and it will work similarly.
+
+This is okay for simple routing. But the API is limited, and if you want to use the full features provided by your Ingress controller, you are stuck with [annotations](https://apisix.apache.org/docs/ingress-controller/concepts/annotations/).
+
+For example, the Kubernetes Ingress API does not provide a schema to configure rewrites. Rewrites are useful when your upstream/backend URL differs from the path configured in your Ingress rule.
+
+APISIX supports this feature, and you have to use custom annotations to leverage it:
+
+```yaml {title="kubernetes-ingress-manifest.yaml"}
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+  name: api-routes
+  annotations:
+    k8s.apisix.apache.org/rewrite-target-regex: "/app/(.*)"
+    k8s.apisix.apache.org/rewrite-target-regex-template: "/$1"
+spec:
+  ingressClassName: apisix
+  rules:
+    - host: local.navendu.me
+      http:
+        paths:
+          - backend:
+              service:
+                name: bare-minimum-api
+                port:
+                  number: 8080
+            path: /app
+            pathType: Prefix
+```
+
+This creates an Ingress resource that configures APISIX to route any requests with the `/app` prefix to the backend with the prefix removed. For example, a request to `/app/version` will be forwarded to `/version`.
+
+Annotations are specific to your choice of an Ingress controller. These "proprietary" extensions limited the scope of portability intended initially with the Ingress API.
+
+## Custom CRDs > Ingress API
+
+Being stuck with annotations also sacrifice the usability of the Ingress controllers.
+
+Controllers therefore solved the limitations of the Ingress API by creating their [own custom resources](https://apisix.apache.org/docs/ingress-controller/references/apisix_route_v2beta3/). The example below shows configuring Ingress to route traffic between two versions of an application using APISIX's custom resource:
+
+```yaml {title="apisix-ingress-manifest.yaml"}
+apiVersion: apisix.apache.org/v2beta3
```

Comment:
https://apisix.apache.org/docs/ingress-controller/concepts/apisix_route/

In APISIX Ingress 1.5 version, we recommend use `apisix.apache.org/v2`. The v2beta3 has been marked as deprecated.